### ISSUE TITLE
TriblerTunnelCommunity.remove_circuit(): Task was destroyed but it is pending!

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#5250]
The following error got caught by the application tester (Windows 7, 64-bit):

```
[PID:2956] 2020-03-26 13:38:13,599 - ERROR <tribler_window:107> root.on_exception(): Traceback (most recent call last):
  File "src\tribler-gui\tribler_gui\event_request_manager.py", line 114, in on_read_data
RuntimeError: Task was destroyed but it is pending!
--CONTEXT--
{'message': 'Task was destroyed but it is pending!', 'task': <Task pending coro=<TriblerTunnelCommunity.remove_circuit() running at src\tribler-core\tribler_core\modules\tunnel\community\triblertunnel_community.py:407> wait_for=<_GatheringFuture pending cb=[<TaskWakeupMethWrapper object at 0x000000001617A6D8>()]> cb=[TaskManager.register_task.<locals>.done_cb() at ipv8\taskmanager.py:105]>}

ERROR:Executor:********** TRIBLER CRASHED **********
ERROR:Executor:Tribler crashed after uptime of 1562.634765625 sec! Stack trace: Traceback (most recent call last):
  File "src\tribler-gui\tribler_gui\event_request_manager.py", line 114, in on_read_data
RuntimeError: Task was destroyed but it is pending!
--CONTEXT--
{'message': 'Task was destroyed but it is pending!', 'task': <Task pending coro=<TriblerTunnelCommunity.remove_circuit() running at src\tribler-core\tribler_core\modules\tunnel\community\triblertunnel_community.py:407> wait_for=<_GatheringFuture pending cb=[<TaskWakeupMethWrapper object at 0x000000001617A6D8>()]> cb=[TaskManager.register_task.<locals>.done_cb() at ipv8\taskmanager.py:105]>}
```

**Comment by qstokkink** [comment]
This is triggering on the `await gather` here (in `remove_circuit()`):

https://github.com/Tribler/tribler/blob/f80d6683cb23f792d111b82cb5bb26b551d75d89/src/tribler-core/tribler_core/modules/tunnel/community/triblertunnel_community.py#L404-L407

**Comment by egbertbouman** [comment]
Do we have any idea whether or not this happened during shutdown? Can I find this on Jenkins somewhere, so that I can investigate this further?

Could possibly be related to the weakref patch in IPv8, since the `TaskManager` would normally prevent this from happening. I'll have a look and see if I can fix it.

**Comment by devos50** [comment]
@egbertbouman Nope, it did not happen during shutdown. There are no log outputs in the application tester on Windows (I think this is somehow related to the setting whether the console log is shown or not). See [here](https://jenkins-ci.tribler.org/job/Deployment-Tests/job/Run_Tribler_Release/job/Run_Win7_64bit/999/console).

**Comment by egbertbouman** [comment]
@devos50 Thanks! I'll have a look.

**Comment by ichorid** [comment]
This stuff is already on Reporter for 7.5.0-rc2.

**Comment by synctext** [comment]
Bug is also confirmed to happen also on Linux 64-bits 18.04

**Comment by egbertbouman** [comment]
I've seen this error twice in a row, but nothing after that. Not a single crash in 7+ hours.

Luckily, it's pretty obvious what's happening here. The `remove_circuit` task gets garbage collected without the `gather` statement having finished. This could happen if there are no references to this task anymore, which is why as was referring to the weakref patch earlier.

Looking at the code, it's clear that `update_torrent` shouldn't be async anyway, especially since waiting for a handle is just asking for problems if you consider that we may never get a handle... So, I'll refactor the `update_torrent` function, and avoid the `gather` statement altogether. Hopefully, this will fix the issue, although the underlying problem seems to be unexpected garbage collection.

**Comment by qstokkink** [comment]
> Luckily, it's pretty obvious what's happening here. The remove_circuit task gets garbage collected without the gather statement having finished. This could happen if there are no references to this task anymore, which is why as was referring to the weakref patch earlier.

Could this not happen with future methods then? Seems like the long-term solution is to go back to a normal dictionary in the TaskManager. We WILL have to fix what was causing the memory leak in the first place then - which is why the weak dict was introduced.

**Comment by ichorid** [comment]
Basically, we want 2 conditions to always hold: 
 1. unfinished Tasks should never be garbage collected.
 2. as soon as a Task is finished, its reference should be removed from the parent TaskManager. 

WeakValueDict guarantees condition 2, but violates condition 1.
Regular Dict guarantees condition 1, but violates condition 2

**Comment by egbertbouman** [comment]
@qstokkink Futures will produce the same error if garbage collected without having finished.
Since I'm having problems reproducing the error I can't really be sure that switching back to a normal dictionary will fix the issue.

**Comment by qstokkink** [comment]
> WeakValueDict guarantees condition 2, but violates condition 1.
Regular Dict guarantees condition 1, but violates condition 2

Given the insane stack-traces that are popping up I think I'm starting to lean toward a regular dict.

@egbertbouman Sorry, miscommunication there. I didn't mean "methods that produce a future", I meant "other methods in the future".

**Comment by ichorid** [comment]
@qstokkink, if you intend to use a regular dict, you'll have to clean up finished Tasks periodically. Though, that would only mask the leakage of those Tasks that got completely stuck.

**Comment by qstokkink** [comment]
> you'll have to clean up finished Tasks periodically. 

We used to have this, but it got removed. We'll reinstate this then:

https://github.com/Tribler/py-ipv8/commit/67be0e25d1b3e525a205279256bca17f6defd2fe#diff-145623a5557fcda4c674f84e896ae105L162-L174

> Though, that would only mask the leakage of those Tasks that got completely stuck.

At least we can log exactly what gets stuck instead of having mysterious errors like this issue pop up.

**Comment by egbertbouman** [comment]
@qstokkink The reason that I removed the periodic cleanup is that it was causing stacktraces to only show up much later (during the cleanup). I replaced it with `done_cb` which should remove the Task directly after it's done (see https://github.com/Tribler/py-ipv8/blob/master/ipv8/taskmanager.py#L106).

So, I guess this means that a lot of tasks never finish. To me, this sounds like a bug, or at least a sign that some tasks need refactoring.

**Comment by qstokkink** [comment]
> So, I guess this means that a lot of tasks never finish. To me, this sounds like a bug, or at least a sign that some tasks need refactoring.

Yes. But we need some way of know what these particular tasks are. I would propose periodically checking the state of the currently registered tasks and logging a warning if any of them stay around for more than.. 10 minutes or so?

**Comment by ichorid** [comment]
@qstokkink , when was the last time you found a bug by looking at a warning in a user's log? :wink:

**Comment by qstokkink** [comment]
@ichorid Well people sent me about 5 messages regarding deserialization warning just a week ago.

**Comment by ichorid** [comment]
@qstokkink , good for you. Seems IPv8 users are more inquiring than Tribler users in general. Of two years here, I never witnessed anyone but developers worry about anything in the logs. :man_shrugging:

**Comment by egbertbouman** [comment]
For testing purposes, I've added a loop to the `TaskManager` that checks for long-running non-interval tasks, and logs a warning if such tasks exist.

It seems that even with a `WeakValueDictionary` we have some memory leakage, although the tasks may get garbage collected eventually. For instance, when the `TorrentChecker` cleans up a session, the cleanup task can get stuck. This is something that's easily fixed but is a clear bug (I'll make a PR for that). So, I think a periodic loop may have some merit regardless of whether we switch back to a real dictionary or not.

**Comment by qstokkink** [comment]
@egbertbouman could you bring that into production?

**Comment by egbertbouman** [comment]
@qstokkink I'll make a PR for it.

**Comment by qstokkink** [comment]
@egbertbouman thanks 😃

**Comment by ichorid** [comment]
@egbertbouman , please don't forget about fixing the original issues that started this thread... :wink:

**Comment by egbertbouman** [comment]
@ichorid Luckily I have you to remind me, thanks!

**Comment by qstokkink** [comment]
https://stackoverflow.com/a/47476998

**Comment by ichorid** [comment]
Closed in #5259
